teamcity: failed tests on master: testrace (proposer evaluated kv)/TestRaceWithBackfill

The following tests appear to have failed:

[#152375](https://teamcity.cockroachdb.com/viewLog.html?buildId=152375):

```
--- FAIL: testrace (proposer evaluated kv)/TestRaceWithBackfill (10.820s)
Race detected!
schema_changer_test.go:426: schema change ALTER TABLE t.test ADD COLUMN x DECIMAL DEFAULT (DECIMAL '1.4') took 1.44562182s
schema_changer_test.go:426: schema change ALTER TABLE t.test DROP pi took 1.528748017s
schema_changer_test.go:426: schema change CREATE UNIQUE INDEX foo ON t.test (v) took 1.019837559s
schema_changer_test.go:426: schema change DROP INDEX t.test@vidx took 1.038285198s
------- Stdout: -------
==================
Read at 0x00c420780518 by goroutine 601:
  github.com/cockroachdb/cockroach/pkg/sql_test.TestRaceWithBackfill.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/schema_changer_test.go:528 +0x4d
  github.com/cockroachdb/cockroach/pkg/sql_test.TestRaceWithBackfill.func3()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/schema_changer_test.go:551 +0xc5
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*indexBackfiller).runChunk.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/indexbackfiller.go:138 +0xc4b
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:482 +0x46
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:515 +0x1ee
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:483 +0x2bc
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*indexBackfiller).runChunk()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/indexbackfiller.go:196 +0x29e
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*indexBackfiller).mainLoop()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/indexbackfiller.go:240 +0x72e
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*indexBackfiller).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/indexbackfiller.go:272 +0x18b

Previous write at 0x00c420780518 by goroutine 389:
  github.com/cockroachdb/cockroach/pkg/sql_test.TestRaceWithBackfill.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/schema_changer_test.go:531 +0x97
  github.com/cockroachdb/cockroach/pkg/sql_test.TestRaceWithBackfill.func3()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/schema_changer_test.go:551 +0xc5
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*indexBackfiller).runChunk.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/indexbackfiller.go:138 +0xc4b
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:482 +0x46
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:515 +0x1ee
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:483 +0x2bc
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*indexBackfiller).runChunk()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/indexbackfiller.go:196 +0x29e
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*indexBackfiller).mainLoop()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/indexbackfiller.go:240 +0x72e
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*indexBackfiller).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/indexbackfiller.go:272 +0x18b

Goroutine 601 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:304 +0x52d
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*flowScheduler).runFlowNow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow_scheduler.go:74 +0x12a
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*flowScheduler).ScheduleFlow.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow_scheduler.go:91 +0xe5
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:239 +0x124
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*flowScheduler).ScheduleFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow_scheduler.go:98 +0xf8
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ServerImpl).SetupFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/server.go:165 +0x1ae
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun._DistSQL_SetupFlow_Handler()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/api.pb.go:256 +0x345
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).processUnaryRPC()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:709 +0xe1e
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:893 +0xf9c
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:469 +0xb8

Goroutine 389 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:304 +0x52d
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*flowScheduler).runFlowNow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow_scheduler.go:74 +0x12a
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*flowScheduler).ScheduleFlow.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow_scheduler.go:91 +0xe5
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:239 +0x124
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*flowScheduler).ScheduleFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow_scheduler.go:98 +0xf8
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ServerImpl).SetupFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/server.go:165 +0x1ae
  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun._DistSQL_SetupFlow_Handler()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/api.pb.go:256 +0x345
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).processUnaryRPC()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:709 +0xe1e
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:893 +0xf9c
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:469 +0xb8
==================
E170216 16:06:41.695832 188779 sql/distsqlrun/server.go:202  [n1] context deadline exceeded
E170216 16:06:41.696094 188774 sql/distsqlrun/server.go:202  [n1] context deadline exceeded
E170216 16:06:41.696584 188576 sql/distsqlrun/server.go:202  [n1] context deadline exceeded
E170216 16:06:41.721511 188716 sql/distsqlrun/server.go:202  [n1] context deadline exceeded

```
Please assign, take a look and update the issue accordingly.

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -43,6 +43,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 )
 
@@ -410,7 +411,7 @@ func runSchemaChangeWithOperations(
 	schemaChange string,
 	maxValue int,
 	keyMultiple int,
-	backfillNotification chan bool,
+	backfillNotification chan struct{},
 ) {
 	tableDesc := sqlbase.GetTableDescriptor(kvDB, "t", "test")
 
@@ -518,13 +519,24 @@ func bulkInsertIntoTable(sqlDB *gosql.DB, maxValue int) error {
 // that run simultaneously.
 func TestRaceWithBackfill(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	var backfillNotification chan bool
+	// protects backfillNotification
+	var mu syncutil.Mutex
+	var backfillNotification chan struct{}
+
 	var partialBackfillDone atomic.Value
 	partialBackfillDone.Store(false)
 	var partialBackfill bool
 	const numNodes, chunkSize, maxValue = 5, 100, 4000
 	params, _ := createTestServerParams()
+	initBackfillNotification := func() chan struct{} {
+		mu.Lock()
+		defer mu.Unlock()
+		backfillNotification = make(chan struct{})
+		return backfillNotification
+	}
 	notifyBackfill := func() {
+		mu.Lock()
+		defer mu.Unlock()
 		if backfillNotification != nil {
 			// Close channel to notify that the backfill has started.
 			close(backfillNotification)
@@ -606,7 +618,6 @@ CREATE UNIQUE INDEX vidx ON t.test (v);
 	// Run some schema changes with operations.
 
 	// Add column.
-	backfillNotification = make(chan bool)
 	runSchemaChangeWithOperations(
 		t,
 		sqlDB,
@@ -614,10 +625,9 @@ CREATE UNIQUE INDEX vidx ON t.test (v);
 		"ALTER TABLE t.test ADD COLUMN x DECIMAL DEFAULT (DECIMAL '1.4')",
 		maxValue,
 		2,
-		backfillNotification)
+		initBackfillNotification())
 
 	// Drop column.
-	backfillNotification = make(chan bool)
 	runSchemaChangeWithOperations(
 		t,
 		sqlDB,
@@ -625,10 +635,9 @@ CREATE UNIQUE INDEX vidx ON t.test (v);
 		"ALTER TABLE t.test DROP pi",
 		maxValue,
 		2,
-		backfillNotification)
+		initBackfillNotification())
 
 	// Add index.
-	backfillNotification = make(chan bool)
 	runSchemaChangeWithOperations(
 		t,
 		sqlDB,
@@ -636,10 +645,9 @@ CREATE UNIQUE INDEX vidx ON t.test (v);
 		"CREATE UNIQUE INDEX foo ON t.test (v)",
 		maxValue,
 		3,
-		backfillNotification)
+		initBackfillNotification())
 
 	// Drop index.
-	backfillNotification = make(chan bool)
 	runSchemaChangeWithOperations(
 		t,
 		sqlDB,
@@ -647,7 +655,7 @@ CREATE UNIQUE INDEX vidx ON t.test (v);
 		"DROP INDEX t.test@vidx",
 		maxValue,
 		2,
-		backfillNotification)
+		initBackfillNotification())
 
 	// Verify that the index foo over v is consistent, and that column x has
 	// been backfilled properly.
@@ -686,8 +694,7 @@ CREATE UNIQUE INDEX vidx ON t.test (v);
 	// This test could be made its own test but is placed here to speed up the
 	// testing.
 
-	notification := make(chan bool)
-	backfillNotification = notification
+	notification := initBackfillNotification()
 	partialBackfill = true
 	// Run the schema change in a separate goroutine.
 	var wg sync.WaitGroup

